### ISSUE TITLE
Require branch name of VCS with possible special chars

### ISSUE THREAD

**Issue by r34ch** [composer/composer#5420]
I am trying to require a specific branch of a vcs package which possibly contains some special chars: `%1-my-i#1-branch-name`

    require: {
        "mypackage/mypackage": "dev-%1-my-i#1-branch-name"
    }

However composer keeps pulling in `dev-master` instead. I have tried escaping certain chars although I don't know which characters composer treats as special chars. I have tried variations of the snippet below

    require: {
        "mypackage/mypackage": "dev-\\%1-my-i\\#1-branch-name"
    }

What am I missing? Or failing that, is it possible to pull a branch via a commit hash only - without specifying the branch name

**Comment by rquadling** [comment]
Environment variables are expanded. I wonder if %1 is attempting to be expanded in some fashion.

If you try with -vvv, do you see anything being substituted?

**Comment by r34ch** [comment]
Doesn't seem like it is bring expanded, just looks like its specifically checking out the latest commit hash of master

    Cloning bc67d8a8db8d6b8db68da60957fc6a18ddc
    Executing command (CWD): git clone --no-checkout ....... && git fetch composer
    Executing command (/path/to/vendor/): git branch -r
    Executing command (/path/to/vendor/): git checkout 'master' --
    Executing command (/path/to/vendor/): git reset --hard 'bc67d8a8db8d6b8db68da60957fc6a18ddc'

**Comment by alcohol** [comment]
Well, `#` is used to denote a commit hash generally (hack, legacy). So it could be that it is stripping that part off. On the other hand though.. not sure. Probably should write a test for this in [composer/semver](/composer/semver).

**Comment by Seldaek** [comment]
IMO this is not something we can fix while retaining the `#hash` feature (see https://getcomposer.org/doc/04-schema.md#package-links) to pin things down to single commits.

Using `#` in branch names is kinda crazy though, as it's bound to create problems in various places.

**Comment by jmalloc** [comment]
I've also encountered this problem as a result of using the [Waffle.io branch naming pattern](https://github.com/waffleio/waffle.io/wiki/FAQs#branch-moving), where hash characters in branch names are part of the usual workflow (this scheme uses the GitHub hash-notation to refer to issues in the branch name).

Obviously, this isn't a problem with Composer specifically, but I figured I'd offer it as one use-case if this ever came again. Perhaps some kind of escape character could be introduced.
